src/logger/fileWriter.c

```diff
@@ -34,47 +35,55 @@
 #define WRITE_SUCCESS  0
 #define WRITE_FAIL     EOF
 
-typedef struct _FileBuffer {
-    char buffer[FILE_BUFFER_SIZE];
-    size_t index;
-} FileBuffer;
-
 static FIL *g_logfile;
 static xQueueHandle g_LoggerMessage_queue;
-static FileBuffer fileBuffer = {"", 0};
-
+static struct ring_buff file_buff;
 
-static void clear_file_buffer() {
-        fileBuffer.index = 0;
-        fileBuffer.buffer[0] = '\0';
+static void error_led(bool on)
+{
+        on ? LED_enable(3) : LED_disable(3);
 }
 
-static int writeFileBuffer()
+static FRESULT flush_file_buffer()
 {
-        unsigned int bw;
-        const int rc = f_write(g_logfile, fileBuffer.buffer,
-                               fileBuffer.index, &bw);
-        clear_file_buffer();
-        return rc;
+        FRESULT res = FR_OK;
+        char tmp[32];
+
+        pr_info("Flushing file buffer\r\n");
+        size_t chars = get_used(&file_buff);
+        while(0 < chars) {
+                if (chars > sizeof(tmp))
+                        chars = sizeof(tmp);
+
+                pr_info_int_msg("Chars is ", chars);
+                get_data(&file_buff, &tmp, chars);
+                if (g_logfile->fs) {
+                        unsigned int bw;
+                        res = f_write(g_logfile, &tmp, chars, &bw);
+                        error_led(FR_OK != res);
+                }
+
+                chars = get_used(&file_buff);
+        }
+
+        pr_info("Flushing file buffer DONE\r\n");
+        return res;
 }
 
-static void appendFileBuffer(const char * data)
+static FRESULT append_file_buffer(const char *data)
 {
-    size_t index = fileBuffer.index;
-    char * buffer = fileBuffer.buffer + index;
-
-    while(*data) {
-        *buffer++ = *data++;
-        index++;
-        if (index >= FILE_BUFFER_SIZE) {
-            *buffer = '\0';
-            writeFileBuffer();
-            index = fileBuffer.index;
-            buffer = fileBuffer.buffer + index;
+        FRESULT res = FR_OK;
+
+        const size_t size = strlen(data);
```

Comment:
The big thing with a special string function is to possibly avoid iterating
the string twice. If you loop once until the null terminator then you'd
avoid the strlen.

Brent Picasso
Autosport Labs
Technology for Race and Street
On Aug 3, 2015 6:45 AM, "Andrew Stiegmann" <notifications@github.com> wrote:

> In src/logger/fileWriter.c
> <https://github.com/autosportlabs/RaceCapture-Pro_firmware/pull/74#discussion_r36085965>
> :
>
> >  {
> > -    size_t index = fileBuffer.index;
> > -    char * buffer = fileBuffer.buffer + index;
> > -
> > -    while(*data) {
> > -        *buffer++ = *data++;
> > -        index++;
> > -        if (index >= FILE_BUFFER_SIZE) {
> > -            *buffer = '\0';
> > -            writeFileBuffer();
> > -            index = fileBuffer.index;
> > -            buffer = fileBuffer.buffer + index;
> > +        FRESULT res = FR_OK;
> > +
> > +        const size_t size = strlen(data);
>
> Of course there could be but I don't see the advantage because strlen is
> only called once per string added. Plus no matter if I call strlen or write
> a put_str method, both have to iterate over the string so nothing is really
> saved IMHO. Plus then i still need to account for partial string puts which
> honestly won't make the code any simpler. So its possible... but I don't
> quite see the benefit. Perhaps I am missing something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/autosportlabs/RaceCapture-Pro_firmware/pull/74/files#r36085965>
> .
>